src/consumer.js

```diff
@@ -0,0 +1,265 @@
+    this.processingMessages = 0;
+    this.idleCallback = null;
+  }
+
+  /**
+   * Create and bind the queue, then start listening.  When this method has
+   * returned, the queue is established and bound to the exchanges given in
+   * the consumer.
+   *
+   * In the public API, this is called automatically by `consume`
+   */
+  async _start() {
+    // first make sure the queue is bound
+    await this.client.withChannel(channel => this._createAndBindQueue(channel));
+
+    // then set up to call _handleConnection on new connections and, if there
+    // is already an active connection, on that as well.
+    this.client.on('connected', this._handleConnection);
+    if (this.client.activeConnection) {
+      this._handleConnection(this.client.activeConnection);
+    }
+  }
+
+  /**
+   * Stop listening.  After this call, no further messages will be consumed
+   * from the queue.  The queue and any bindings will remain configured on the
+   * server.  This method will return after any pending consumers have
+   * completed (ack or nack).
+   */
+  async stop() {
+    if (!this.running) {
+      return;
+    }
+    this.running = false;
+
+    return this._shutdown();
+  }
+
+  /**
+   * Shut down this listener and channel, without setting this.running to false
+   */
+  async _shutdown() {
+    const {channel, consumerTag} = this;
+
+    if (channel && consumerTag) {
+      this.consumerTag = null;
+      await channel.cancel(consumerTag);
+    }
+
+    // if messages are being processed, arrange to continue when they
+    // are all handled
+    if (this.processingMessages > 0) {
+      await new Promise(resolved => {
+        this.idleCallback = resolved;
+      });
+    }
+
+    if (channel) {
+      this.channel = null;
+      await channel.close();
+    }
+  }
+
+  async _createAndBindQueue(channel) {
+    const queueName = this.client.fullObjectName(
+      'queue',
+      this.queueName || `exclusive/${slugid.v4()}`);
+    await channel.assertQueue(queueName, {
+      exclusive: this.exclusiveQueue,
+      durable: !this.exclusiveQueue,
+      autoDelete: this.exclusiveQueue,
+      maxLength: this.options.maxLength,
+    });
+
+    for (let {exchange, routingKeyPattern} of this.bindings) {
+      await channel.bindQueue(queueName, exchange, routingKeyPattern);
+    }
+
+    return queueName;
+  }
+
+  /**
+   * Handle a new connection to the pulse server, re-declaring everything
+   * and setting up the consumer.
+   */
+  async _handleConnection(conn) {
+    try {
+      if (!this.running) {
+        return;
+      }
+
+      const amqp = conn.amqp;
+      const channel = await amqp.createChannel();
+      await channel.prefetch(this.options.prefetch);
+      const queueName = await await this._createAndBindQueue(channel);
+      this.channel = channel;
+
+      // consider any errors on the channel to be potentially fatal to the
+      // connection (better safe than sorry)
+      channel.on('error', () => conn.failed());
```

Comment:
No, channel errors are not connection errors.

RabbitMQ is not insane :)

If we have an error at channel, the channel is dead, typically because you made a mistake... But typically, it won't kill the connection, if it does having `.on('error', ` on the connection should be sufficient.

Reply:
I added this at your suggestion, iirc :)

Reply:
differentiating between connection and channel errors by experimenting will make things clear.If channel error is due to some user's invalid data through `channel.someMethod` then repeating `createChannel` may not be right or there should be no. of reattempt limits.